main/lsp/UndoState.cc

```diff
@@ -13,28 +13,22 @@ UndoState::UndoState(const LSPConfiguration &config, unique_ptr<core::GlobalStat
     : config(config), evictedGs(move(evictedGs)), evictedIndexedFinalGS(std::move(evictedIndexedFinalGS)),
       evictedFilesThatHaveErrors(move(evictedFilesThatHaveErrors)) {}
 
-void UndoState::recordEvictedState(ast::ParsedFile evictedIndexTree, core::FileHash evictedFileHash) {
+void UndoState::recordEvictedState(ast::ParsedFile evictedIndexTree) {
     const auto id = evictedIndexTree.file.id();
     // The first time a file gets evicted, it's an index tree from the old global state.
     // Subsequent times it is evicting old index trees from the new global state, and we don't care.
     // Also, ignore updates to new files (id >= size of file table)
     if (id < evictedGs->getFiles().size() && !evictedIndexed.contains(id)) {
         evictedIndexed[id] = move(evictedIndexTree);
-        // file hashes should be kept in-sync with indexed.
```

Comment:
This is subtle, but we don't have to record evicted files in `UndoState` and explicitly "restore" them.

Files (and, now, their hashes!) are stored in `GlobalState`. `UndoState` contains the old `GlobalState`, which contains the old files -- so reinstating it after cancelation is sufficient to bring back the old files (and their hashes).